### PR TITLE
ab2d 973 introduce failure threshold for the worker

### PR DESCRIPTION
**Pull request by spathiyil**
introduce failure threshold for the worker

### Summary

Given a job, the worker processes all the patients in the contract. Introduce a failure threshold so that the batch is not stopped on the first failure. Instead when a failure occurs, check against an externally configurable failure threshold and if the number of errors crosses the threshold, only then stop processing the job.

### Checklist

- [x] Tests and linting pass

### Security
N/A

### Additional JIRA Tickets (optional)

N/A

### CODE REVIEW

File: worker/src/main/java/gov/cms/ab2d/worker/processor/domainmodel/ProgressTracker.java [hunk 1]
```diff
@@ -57,5 +64,9 @@ public int getPercentageCompleted() {
         return percentCompleted;
     }
 
+    public boolean isErrorCountBelowThreshold() {
+        return (failureCount * 100) / processedCount < failureThreshold;
```
Comment (dkrylovsb) [inline worker/src/main/java/gov/cms/ab2d/worker/processor/domainmodel/ProgressTracker.java]: The failure threshold should be calculated against the total number of patients in the contract.

Comment (spathiyil) [reply worker/src/main/java/gov/cms/ab2d/worker/processor/domainmodel/ProgressTracker.java]: done.

File: worker/src/main/resources/application.properties [hunk 2]
```diff
@@ -19,6 +19,9 @@ cancellation.check.frequency=10
 report.progress.db.frequency=100
 report.progress.log.frequency=10000
 
+## fail the job if >= 10% of the records fail
+failure.threshold=10
```
Comment (dkrylovsb) [inline worker/src/main/resources/application.properties]: Let's start with a conservative default of 1%. Still a large number for large contracts.

File: worker/src/main/java/gov/cms/ab2d/worker/processor/JobProcessorImpl.java [hunk 3]
```diff
@@ -367,6 +361,20 @@ private void processHandles(List<Future<Void>> futureHandles, ProgressTracker pr
...
+    private void analyzeException(List<Future<Void>> futureHandles, ProgressTracker progressTracker, Exception e) {
+        progressTracker.incrementFailureCount();
+        progressTracker.incrementProcessedCount();
+        if (progressTracker.isErrorCountBelowThreshold()) {
+            final Throwable rootCause = ExceptionUtils.getRootCause(e);
+            log.error("exception while processing patient {}", rootCause.getMessage(), rootCause);
+            // log exception, but continue processing job as errorCount is below threshold
+        } else {
+            cancelFuturesInQueue(futureHandles);
+            log.error("{} out of {} records failed. Stopping job", progressTracker.getFailureCount(), progressTracker.getProcessedCount());
```
Comment (dkrylovsb) [inline worker/src/main/java/gov/cms/ab2d/worker/processor/JobProcessorImpl.java]: I guess now the 2nd parameter should be `totalCount`.

File: worker/src/test/resources/application.properties [hunk 4]
```diff
@@ -22,6 +22,9 @@ cancellation.check.frequency=2
 report.progress.db.frequency=10
 report.progress.log.frequency=10
 
+## fail the job if >= 10% of the records fail
+failure.threshold=10
```
Comment (dkrylovsb) [inline worker/src/test/resources/application.properties]: Fine to keep `10` in tests of course

File: worker/src/main/java/gov/cms/ab2d/worker/processor/JobProcessorImpl.java (hunk 3 above)
Comment (spathiyil) [reply worker/src/main/java/gov/cms/ab2d/worker/processor/JobProcessorImpl.java]: changed.

**Pull request merged**